rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,105 @@
+  name: folks
+  namespace: teleport
+instanceName: some-instance
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+instanceName: some-instance
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport. Every custom resource must have the `instanceName` field useful to look up an instance resource in order to know what address to connect to. Instance resource and its associated secret must reside in the same namespace as the custom resource.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields, `instanceName` in particular. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### Identity definitions
+
+To run some plugin in the cluster, the credentials are required. Normally, one can get the credentials credentials using `tctl auth sign` command or using `tsh login`. To automate the credentials generation, the `Identity` resource is introduced. The identity definition looks like this:
+
+```yaml
+apiVersion: control.goteleport.com/v1
+kind: Identity
+metadata:
+  name: access-slack
+spec:
+  username: access-slack
+  secretName: access-slack-secret
+```
+
+Here, the `username` is a Teleport user name, and `secretName` is a name of a secret resource where the manager will write credentials. Once the `Identity` object is created in the Kubernetes, the secret object with credentials contents will be written in the same namespace where the identity object resides.
+
+### Limit the scope
+
+When running the manager as a sidecar container, the question arises: what objects is this particular instance of the manager responsible for and what objects not? There could be multiple Teleport deployments in the same cluster, and each deployment might have its manager sidecar container. So the question is how to limit the scope of objects observed.
+
+The first option is to limit by the namespace, and the second is to limit by label selectors. Both of them must be present in the manager to be very specific on what Kubernetes objects to watch.
```

Comment:
I wonder, why not just have the operator monitor resources in the same namespace it's running in? Then there would no be any additional configuration. And it will work for Teleport Cloud use-case as well.

Reply:
@r0mant Yes but label selectors configuration is optional. If it's missing the selectors won't be applied.

I brought the label selectors to serve the helm chart use case — to explicitely limit the resources the manager handles only to those that were created by Helm. The helm chart uses label selectors everywhere so I just wanted to follow the practice.
Don't you think it's a right idea?

Reply:
Limiting by namespace seems to be straight-forward enough. I don't know if we need the label selectors, to be honest.

Reply:
Okay lets stay with just namespaces for now.

Reply:
Yeah I agree with just using namespaces. Let's please clarify this in the RFD so it's explicit what's going to be implemented and what's out of scope.